File: src/base32hex.test.ts
import {
	describe,
	expect,
	test,
} from "bun:test";

import {
	decode,
	encode,
} from "./base32hex.js";

const base32hexTestCases = [
	{
		name: "empty string",
		input: new Uint8Array([]),
		expectedEncoded: "",
	},
	{
		name: "single byte",
		input: new Uint8Array([102]),
		expectedEncoded: "CO======",
	},
	{
		name: "two bytes",
		input: new Uint8Array([102, 111]),
		expectedEncoded: "CPNG====",
	},
	{
		name: "three bytes",
		input: new Uint8Array([102, 111, 111]),
		expectedEncoded: "CPNMU===",
	},
	{
		name: "four bytes",
		input: new Uint8Array([102, 111, 111, 98]),
		expectedEncoded: "CPNMUOG=",
	},
	{
		name: "five bytes",
		input: new Uint8Array([102, 111, 111, 98, 97]),
		expectedEncoded: "CPNMUOJ1",
	},
	{
		name: "six bytes",
		input: new Uint8Array([102, 111, 111, 98, 97, 114]),
		expectedEncoded: "CPNMUOJ1E8======",
	},
];

const base32hexNoPaddingTestCases = [
	{ name: "single byte no padding", input: new Uint8Array([102]), expectedEncoded: "CO" },
	{ name: "two bytes no padding", input: new Uint8Array([102, 111]), expectedEncoded: "CPNG" },
	{
		name: "three bytes no padding",
		input: new Uint8Array([102, 111, 111]),
		expectedEncoded: "CPNMU",
	},
	{
		name: "four bytes no padding",
		input: new Uint8Array([102, 111, 111, 98]),
		expectedEncoded: "CPNMUOG",
	},
	{
		name: "five bytes no padding",
		input: new Uint8Array([102, 111, 111, 98, 97]),
		expectedEncoded: "CPNMUOJ1",
	},
	{
		name: "six bytes no padding",
		input: new Uint8Array([102, 111, 111, 98, 97, 114]),
		expectedEncoded: "CPNMUOJ1E8",
	},
];

describe("base32hex encoding", () => {
	test.each(base32hexTestCases)("$name", ({ input, expectedEncoded }) => {
		const output = encode(input);
		expect(output).toBe(expectedEncoded);
	});

	test.each(base32hexNoPaddingTestCases)("$name", ({ input, expectedEncoded }) => {
		const withPadding = false;
		const output = encode(input, withPadding);
		expect(output).toBe(expectedEncoded);
	});
});

const base32hexDecodeTestCases = [
	{
		name: "decode empty string",
		input: "",
		expected: new Uint8Array([]),
	},
	{
		name: "decode single byte",
		input: "CO======",
		expected: new Uint8Array([102]),
	},
	{
		name: "decode two bytes",
		input: "CPNG====",
		expected: new Uint8Array([102, 111]),
	},
	{
		name: "decode three bytes",
		input: "CPNMU===",
		expected: new Uint8Array([102, 111, 111]),
	},
	{
		name: "decode four bytes",
		input: "CPNMUOG=",
		expected: new Uint8Array([102, 111, 111, 98]),
	},
	{
		name: "decode five bytes",
		input: "CPNMUOJ1",
		expected: new Uint8Array([102, 111, 111, 98, 97]),
	},
	{
		name: "decode six bytes",
		input: "CPNMUOJ1E8======",
		expected: new Uint8Array([102, 111, 111, 98, 97, 114]),
	},
];

const base32hexLooseModeTestCases = [
	{
		name: "decode with lowercase",
		input: "cpnmu===",
		expected: new Uint8Array([102, 111, 111]),
	},
	{
		name: "decode with mixed case",
		input: "CpNmU===",
		expected: new Uint8Array([102, 111, 111]),
	},
	{
		name: "decode without padding",
		input: "CPNMU",
		expected: new Uint8Array([102, 111, 111]),
	},
];

describe("base32hex decoding", () => {
	test.each(base32hexDecodeTestCases)("$name", ({ input, expected }) => {
		const output = decode(input);
		expect(output).toEqual(expected);
	});

	test.each(base32hexLooseModeTestCases)("$name", ({ input, expected }) => {
		const looseMode = true;
		const output = decode(input, looseMode);
		expect(output).toEqual(expected);
	});

	test("invalid character", () => {
		const invalidInput = "CPNM!";
		expect(() => decode(invalidInput)).toThrow("Invalid base32hex string");
	});
});

File: go/pkg/bysquare/base32hex_test.go
package bysquare

import (
	"bytes"
	"testing"
)

func TestEncodeBase32Hex(t *testing.T) {
	testCases := []struct {
		name     string
		input    []byte
		expected string
	}{
		{
			name:     "empty bytes",
			input:    []byte{},
			expected: "",
		},
		{
			name:     "single byte",
			input:    []byte{102},
			expected: "CO======",
		},
		{
			name:     "two bytes",
			input:    []byte{102, 111},
			expected: "CPNG====",
		},
		{
			name:     "three bytes",
			input:    []byte{102, 111, 111},
			expected: "CPNMU===",
		},
		{
			name:     "four bytes",
			input:    []byte{102, 111, 111, 98},
			expected: "CPNMUOG=",
		},
		{
			name:     "five bytes",
			input:    []byte{102, 111, 111, 98, 97},
			expected: "CPNMUOJ1",
		},
		{
			name:     "six bytes",
			input:    []byte{102, 111, 111, 98, 97, 114},
			expected: "CPNMUOJ1E8======",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			result := encodeBase32Hex(tc.input, true)
			if result != tc.expected {
				t.Errorf("expected %q, got %q", tc.expected, result)
			}
		})
	}
}

func TestEncodeBase32HexNoPadding(t *testing.T) {
	testCases := []struct {
		name     string
		input    []byte
		expected string
	}{
		{
			name:     "single byte no padding",
			input:    []byte{102},
			expected: "CO",
		},
		{
			name:     "two bytes no padding",
			input:    []byte{102, 111},
			expected: "CPNG",
		},
		{
			name:     "three bytes no padding",
			input:    []byte{102, 111, 111},
			expected: "CPNMU",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			result := encodeBase32Hex(tc.input, false)
			if result != tc.expected {
				t.Errorf("expected %q, got %q", tc.expected, result)
			}
		})
	}
}

func TestDecodeBase32Hex(t *testing.T) {
	testCases := []struct {
		name     string
		input    string
		expected []byte
		hasError bool
	}{
		{
			name:     "empty string",
			input:    "",
			expected: []byte{},
			hasError: false,
		},
		{
			name:     "single byte with padding",
			input:    "CO======",
			expected: []byte{102},
			hasError: false,
		},
		{
			name:     "two bytes with padding",
			input:    "CPNG====",
			expected: []byte{102, 111},
			hasError: false,
		},
		{
			name:     "three bytes with padding",
			input:    "CPNMU===",
			expected: []byte{102, 111, 111},
			hasError: false,
		},
		{
			name:     "four bytes with padding",
			input:    "CPNMUOG=",
			expected: []byte{102, 111, 111, 98},
			hasError: false,
		},
		{
			name:     "five bytes no padding needed",
			input:    "CPNMUOJ1",
			expected: []byte{102, 111, 111, 98, 97},
			hasError: false,
		},
		{
			name:     "six bytes with padding",
			input:    "CPNMUOJ1E8======",
			expected: []byte{102, 111, 111, 98, 97, 114},
			hasError: false,
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			result, err := decodeBase32Hex(tc.input, false)
			if tc.hasError {
				if err == nil {
					t.Error("expected error, got nil")
				}
				return
			}
			if err != nil {
				t.Errorf("unexpected error: %v", err)
				return
			}
			if !bytes.Equal(result, tc.expected) {
				t.Errorf("expected %v, got %v", tc.expected, result)
			}
		})
	}
}

func TestDecodeBase32HexLooseMode(t *testing.T) {
	testCases := []struct {
		name     string
		input    string
		expected []byte
		hasError bool
	}{
		{
			name:     "no padding in loose mode",
			input:    "CO",
			expected: []byte{102},
			hasError: false,
		},
		{
			name:     "partial padding in loose mode",
			input:    "CPNG",
			expected: []byte{102, 111},
			hasError: false,
		},
		{
			name:     "lowercase letters in loose mode",
			input:    "cpng",
			expected: []byte{102, 111},
			hasError: false,
		},
		{
			name:     "mixed case in loose mode",
			input:    "CpNg",
			expected: []byte{102, 111},
			hasError: false,
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			result, err := decodeBase32Hex(tc.input, true)
			if tc.hasError {
				if err == nil {
					t.Error("expected error, got nil")
				}
				return
			}
			if err != nil {
				t.Errorf("unexpected error: %v", err)
				return
			}
			if !bytes.Equal(result, tc.expected) {
				t.Errorf("expected %v, got %v", tc.expected, result)
			}
		})
	}
}

func TestBase32HexRoundTrip(t *testing.T) {
	testCases := [][]byte{
		{},
		{0},
		{255},
		{0, 1, 2, 3, 4, 5},
		{102, 111, 111, 98, 97, 114},
		[]byte("Hello, World!"),
		[]byte("The quick brown fox jumps over the lazy dog"),
	}

	for _, original := range testCases {
		encoded := encodeBase32Hex(original, true)
		decoded, err := decodeBase32Hex(encoded, false)
		if err != nil {
			t.Errorf("decode error for %v: %v", original, err)
			continue
		}
		if !bytes.Equal(decoded, original) {
			t.Errorf("round trip failed: original=%v, encoded=%q, decoded=%v",
				original, encoded, decoded)
		}
	}
}

func TestDecodeBase32HexInvalidInput(t *testing.T) {
	testCases := []struct {
		name  string
		input string
	}{
		{
			name:  "invalid character",
			input: "ABC@",
		},
		{
			name:  "invalid character in middle",
			input: "AB$C",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			_, err := decodeBase32Hex(tc.input, false)
			if err == nil {
				t.Error("expected error for invalid input, got nil")
			}
		})
	}
}
